### PR TITLE
Fix/expose invariant exception

### PR DESCRIPTION
**Pull request by bp-kelley**
Converts Invar::Invariant exception into a proper python exception.

Note that in python consoles, this looks like a double log, one is actually going to stderr and one is part of the exception object.  The exception object has been sanitized:

 * From below: /Users/kellebr5/devel/github/bpkelley/rdkit-fixes/ has been removed since this is build info that isn't necessary.
 * The current RDKIT and Boost versions are recorded which may be useful for bug reports

This should have no effect on existing code as we still throw a runtime error, we only add info to the exception.

```
>>> from rdkit import Chem
>>>m = Chem.MolFromSmiles("C")
>>> m.GetAtomWithIdx(3)
[16:04:34] 

****
Range Error
idx
Violation occurred on line 155 in file /Users/kellebr5/devel/github/bpkelley/rdkit-fixes/Code/GraphMol/ROMol.cpp
Failed Expression: 3 <= 0
****

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Range Error
	idx
	Violation occurred on line 155 in file Code/GraphMol/ROMol.cpp
	Failed Expression: 3 <= 0
	RDKIT: 2016.03.1.dev1
	BOOST: 1_58
```

### CODE REVIEW

File: Code/GraphMol/Wrap/rdchem.cpp [hunk 1]
```diff
@@ -59,7 +59,10 @@ BOOST_PYTHON_MODULE(rdchem) {
       &translate_value_error);
   python::register_exception_translator<RDKit::MolSanitizeException>(
       &rdSanitExceptionTranslator);
-
+#if INVARIANT_EXCEPTION_METHOD
```
Comment (greglandrum) [inline Code/GraphMol/Wrap/rdchem.cpp]: This belongs in rdbase (i.e. RDBoost/Wrap.RDBase.cpp), not here. That way the translator will be registered even if rdchem isn't imported.

I know we have exception translators for ValueError and IndexError registered in multiple places too, but I don't think it's necessary anymore (or maybe ever) and I will put it on the "to be cleaned up" list.

File: Code/RDGeneral/Invariant.cpp [hunk 2]
```diff
@@ -46,4 +47,35 @@ std::string Invariant::toString() const {
...
+  std::size_t pos = filename.find("Code"); // strip out build directory info
+  if (pos != std::string::npos) {
+    filename = filename.substr(pos);
+  }
+  
+  std::string stringRep = this->prefix_d + "\n\t" + this->getMessage() +
+      "\n\tViolation occurred on line " + line + " in file " +
+      filename + "\n\tFailed Expression: " +
+      this->getExpression() + "\n\t" +
+      "RDKIT: " + RDKit::rdkitVersion + "\n\t" +
```
Comment (greglandrum) [inline Code/RDGeneral/Invariant.cpp]: This, along with the boost version, is a nice addition.

**Comment by greglandrum** [comment]
With the move of the handler to rdbase, this is ready to go. It's really going to help the jupyter notebook users. :-)

**Comment by bp-kelley** [comment]
Agreed, although this PR has the same effect:

https://github.com/rdkit/rdkit/pull/739

And in some ways is probably more important due to logging errors in Smiles/Smarts parsing that would otherwise be hard to catch.

File: Code/RDGeneral/Invariant.cpp (hunk 2 above)
Comment (bp-kelley) [inline Code/RDGeneral/Invariant.cpp]: I could just put this in toString and have just one function.  I kind of like having the full path for dev purposes which is perhaps a bit selfish :)

**Comment by bp-kelley** [comment]
This file ending stuff is pretty painful.

**Comment by greglandrum** [comment]
Yeah, that change made a complete mess. I apologize for the blunt-hammer fix that I applied to get that file cleaned up, but I wanted to just rip off the band-aid.

**Pull request merged**